probtorch/objectives/marginal.py

```diff
@@ -0,0 +1,139 @@
+                     size_average=size_average, reduce=reduce) -
+            kl(q, p, sample_dim, batch_dim, log_weights, beta,
+               size_average=size_average, reduce=reduce, N=N) +
+            (beta[4] + alpha) * ml(q, sample_dim, batch_dim, log_weights,
+                                   size_average=size_average, reduce=reduce))
+
+
+def kl(q, p, sample_dim=None, batch_dim=None, log_weights=None, beta=(1.0, 1.0, 1.0, 1.0, 1.0),
+         size_average=True, reduce=True, N=None):
+    r"""
+    Computes a Monte Carlo estimate of the unnormalized KL divergence
+    described for variable z.
+    .. math::
+       E_{q(z | x, y)} \left[ \log \frac{q(z | x, y)}{p(z)} \right]
+       \simeq
+       \frac{1}{S} \frac{1}{B} \sum_{s=1}^S \sum_{b=1}^B
+       \left[\beta_{1}  \log \frac{q_{avg}(z^{(s,b)})}{\prod_{d=1}^{D}q_{avg}(z_{d}^{(s,b)})} + \\
+             \beta_{2} \log \prod_{d=1}^{D} \frac{q_{avg}(z_{d}^{(s,b)})}{p(z_{d}^{(s,b)})} +
+             \beta_{3} \log \frac{q(z^{(s,b)} | x^{(b)}, y^{(s,b)})}{q_{avg}(z^{(s,b)})}
+             \beta_{4} \log \frac{q(y^{(b)} | x^{(b)})}{p(y^(b))}
+       \right]
+
+    The sets of variables :math:`x`, :math:`y` and :math:`z` refer to:
+        :math:`x`: The set of conditioned nodes that are present in `p` but
+        are not present in `q`.
+        :math:`y`: The set of conditioned nodes in `q`, which may or may
+        not also be present in `q`.
+        :math:`z`: The set of sampled nodes present in both `q` and `p`.
+    Importance sampling is used to approximate the expectation over
+    :math:`q(z| x,y)`.
+    Arguments:
+        q(:obj:`Trace`): The encoder trace.
+        p(:obj:`Trace`): The decoder trace.
+        sample_dim(int, optional): The dimension containing individual samples.
+        batch_dim(int, optional): The dimension containing batch items.
+        log_weights(:obj:`Variable` or number, optional): Log weights for
+            samples. Calculated when not specified.
+        beta(tuple of int): Containing coefficients for total correlation,
+            kl to the prior, and mutual information
+        size_average (bool, optional): By default, the objective is averaged
+            over items in the minibatch. When set to false, the objective is
+            instead summed over the minibatch.
+        reduce (bool, optional): By default, the objective is averaged or
+           summed over items in the minibatch. When reduce is False, losses
+           are returned without averaging or summation.
+    """
+    y = q.conditioned()
+    if log_weights is None:
+        log_weights = q.log_joint(sample_dim, batch_dim, y)
+    log_qy = log_weights
+    log_py = p.log_joint(sample_dim, batch_dim, y)
+    z = [n for n in q.sampled() if n in p]
+    _,_, log_avg_pzd_prod = p.log_batch_marginal(sample_dim, batch_dim, z, N=N)
+    log_joint_avg_qz, log_avg_qz, log_avg_qzd_prod = q.log_batch_marginal(sample_dim, batch_dim, z, N=N)
```

Comment:
If this just computes the pairwise probabilities, then I think you can calculate the other terms here in the objective. This also means that you don't have to pass the dataset size to `log_batch_marginal`.